File: thsound-code/code/android/TasksProject/Tasks/src/main/java/com/programmingsound/tasks/audio/PdInterface.java
/***
 * Excerpted from "Programming Sound with Pure Data",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material, 
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose. 
 * Visit http://www.pragmaticprogrammer.com/titles/thsound for more book information.
***/
package com.programmingsound.tasks.audio;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

import org.puredata.android.io.AudioParameters;
import org.puredata.android.io.PdAudio;
import org.puredata.core.PdBase;
import org.puredata.core.utils.IoUtils;

import com.programmingsound.tasks.R;

import android.content.Context;
import android.os.Handler;

public class PdInterface {
  private static final int MIN_SAMPLE_RATE        = 44100;
  private static final String MODULATOR_1_TUNE 	  = "mod1_tune";
  private static final String MODULATOR_2_TUNE 	  = "mod2_tune";
  private static final String MODULATOR_1_DEPTH 	= "mod1_depth";
  private static final String MODULATOR_2_DEPTH 	= "mod2_depth";
  private static final String MIDINOTE1 			    = "midinote1";
  private static final String MIDINOTE2 			    = "midinote2";
  private static final long NOTE_DELAY 			      = 120;
  private static final int[] scaleDegrees         = {
    0, 2, 4, 5, 7, 9, 11, 12, 14, 16 
  };
  private static final int ROOT_MIDI_NOTE         = 60;

  private static PdInterface ourInstance = new PdInterface();
  private Context context;
  private Handler handler;
  private int currentRootNote;
  private int currentSecondNote;
  private int currentDegree;
  private int currentOctave;

  public static PdInterface getInstance() {
    return ourInstance;
  }

  private PdInterface() { }

  public void initialize(Context context) {
    this.context = context;
    this.handler = new Handler();
    try {
      initializePd();
      resetScaleDegree();
    } catch (IOException e) {
      // Unable to initialize audio. Ignore silently as this is not fatal to the app
      e.printStackTrace();
    }
  }

  public void destroy() {
    cleanup();
  }

  public void playNewTaskCue() {
    PdBase.sendFloat(MIDINOTE1, ROOT_MIDI_NOTE);
    resetScaleDegree();
  }

  public void playCompleteTaskCue() {
    final float third = currentSecondNote;
    PdBase.sendFloat(MIDINOTE1, currentRootNote);
    handler.postDelayed(new Runnable() {
      @Override public void run() {
        PdBase.sendFloat(MIDINOTE2, third);
      }
    }, NOTE_DELAY);
    incrementScaleDegree();
  }

  public void playClearTasksCue() {
    final int root = ROOT_MIDI_NOTE;
    final int third = ROOT_MIDI_NOTE + scaleDegrees[2];
    PdBase.sendFloat(MIDINOTE1, root);
    PdBase.sendFloat(MIDINOTE2, third);
    handler.postDelayed(new Runnable() {
      @Override public void run() {
        PdBase.sendFloat(MIDINOTE1, root + 12);
        PdBase.sendFloat(MIDINOTE2, third + 12);
      }
    }, NOTE_DELAY);
    resetScaleDegree();
  }
  
  protected void resetScaleDegree() {
    currentRootNote = ROOT_MIDI_NOTE;
    int secondDegree = scaleDegrees[2];
    currentSecondNote = ROOT_MIDI_NOTE + secondDegree;
    currentDegree = 0;
    currentOctave = 0;
  }
  
  protected void incrementScaleDegree() {
    currentDegree++;
    if (currentDegree == 7) {
      currentDegree = 0;
      currentOctave++;
    }
    int rootDegree = scaleDegrees[currentDegree];
    int nextDegree = scaleDegrees[currentDegree + 2];
    currentRootNote = ROOT_MIDI_NOTE + (12 * currentOctave) + rootDegree;
    currentSecondNote = ROOT_MIDI_NOTE + (12 * currentOctave) + nextDegree;
  }

  private void initializePd() throws IOException {
    AudioParameters.init(context);
    int sampleRate = Math.max(MIN_SAMPLE_RATE, AudioParameters.suggestSampleRate());
    int outChannels = AudioParameters.suggestOutputChannels();
    PdAudio.initAudio(sampleRate, 0, outChannels, 1, true);
    File dir = context.getFilesDir();
    File patchFile = new File(dir, "task_tones.pd");
    InputStream patchStream = context.getResources().openRawResource(R.raw.patches);
    IoUtils.extractZipResource(patchStream, dir, true);

    PdBase.openPatch(patchFile.getAbsolutePath());
    PdBase.sendFloat(MODULATOR_1_TUNE, 0);
    PdBase.sendFloat(MODULATOR_1_DEPTH, 0);
    PdBase.sendFloat(MODULATOR_2_TUNE, 0.5f);
    PdBase.sendFloat(MODULATOR_2_DEPTH, 10000);
  }

  private void cleanup() {
    PdAudio.release();
    PdBase.release();
  }

}

File: thsound-code/code/ios/Tasks/Tasks/PdInterface.m
/***
 * Excerpted from "Programming Sound with Pure Data",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material, 
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose. 
 * Visit http://www.pragmaticprogrammer.com/titles/thsound for more book information.
***/
//
//  AudioManager.m
//  Tasks
//
//  Created by Tony Hillerson on 8/18/13.
//  Copyright (c) 2013 Programming Sound. All rights reserved.
//

#import "PdInterface.h"
#import "PdDispatcher.h"
#import "PdBase.h"

#define kModulator1Tune @"mod1_tune"
#define kModulator1Depth @"mod1_depth"
#define kModulator2Tune @"mod2_tune"
#define kModulator2Depth @"mod2_depth"
#define kMidiNote1 @"midinote1"
#define kMidiNote2 @"midinote2"
#define kNoteDelay .12
#define kRootMidiNote 60.

@interface PdInterface () {
    void *patch;
    int currentOctave;
    int currentDegree;
    int currentRootNote;
    int currentSecondNote;
    NSArray *scaleDegrees;
}

@property(nonatomic, strong) PdDispatcher *pdDispatcher;

@end

@implementation PdInterface

- (id) init {
    self = [super init];
    if (self) {
        scaleDegrees = @[@0, @2, @4, @5, @7, @9, @11, @12, @14, @16];
        self.pdDispatcher = [[PdDispatcher alloc] init];
        [PdBase setDelegate:self.pdDispatcher];
        patch = [PdBase openFile:@"task_tones.pd"
                            path:[[NSBundle mainBundle] resourcePath]];
        // Initialize the FM operator settings. Could be read from config
        [PdBase sendFloat:8. toReceiver:kModulator1Tune];
        [PdBase sendFloat:5000. toReceiver:kModulator1Depth];
        [PdBase sendFloat:2.5 toReceiver:kModulator2Depth];
        [PdBase sendFloat:12000. toReceiver:kModulator2Tune];
        [self resetScaleDegree];
    }
    return self;
}

- (void) resetScaleDegree {
    currentRootNote = kRootMidiNote;
    int secondDegree = [scaleDegrees[2] integerValue];
    currentSecondNote = kRootMidiNote + secondDegree;
    currentDegree = 0;
    currentOctave = 0;
}

- (void) incrementScaleDegree {
    currentDegree = currentDegree + 1;
    if (currentDegree == 7) {
        currentDegree = 0;
        currentOctave = currentOctave + 1;
    }
    int rootDegree = [scaleDegrees[currentDegree] integerValue];
    int nextDegree = [scaleDegrees[currentDegree + 2] integerValue];
    currentRootNote = kRootMidiNote + (12 * currentOctave) + rootDegree;
    currentSecondNote = kRootMidiNote + (12 * currentOctave) + nextDegree;
}

- (void) playTaskCreatedCue {
    [PdBase sendFloat:kRootMidiNote toReceiver:kMidiNote1];
    [self resetScaleDegree];
}

- (void) playTaskCompletionCue {
    float secondNote = currentSecondNote;
    [PdBase sendFloat:currentRootNote toReceiver:kMidiNote1];
    dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW,
                                            (int64_t)(kNoteDelay * NSEC_PER_SEC));
    dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
        [PdBase sendFloat:secondNote toReceiver:kMidiNote2];
    });
    [self incrementScaleDegree];
}

- (void) playTasksClearedCue {
    float root = kRootMidiNote;
    float third = kRootMidiNote + [scaleDegrees[2] integerValue];
    [PdBase sendFloat:root toReceiver:kMidiNote1];
    [PdBase sendFloat:third toReceiver:kMidiNote2];
    dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW,
                                            (int64_t)(kNoteDelay * NSEC_PER_SEC));
    dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
        [PdBase sendFloat:root + 12 toReceiver:kMidiNote1];
        [PdBase sendFloat:third + 12 toReceiver:kMidiNote2];
    });
    [self resetScaleDegree];
}

@end
